### ISSUE TITLE
segfault with --hwdec=auto (really all --hwdec options except no)

### ISSUE THREAD

**Issue by afarley** [mpv-player/mpv#9216]
### Important Information

Provide following Information:
- mpv version: 0.33.0-270-g24357cb7b5
- Linux Distribution and Version: Linux 5.14.2-1-MANJARO x86_64 21.1.2 Pahvo
- Source of the mpv binary: compiled from source
- If known which version of mpv introduced the problem: from git between September 1, 2021 and September 10, 2021
- Window Manager and version: mutter 40.4-1
- GPU driver and version: Nvidia 470.63.01-7

### Reproduction steps

play video with any of the --hwdec= options (except --hwdec=no which does not segfault)

### Expected behavior
video would play

### Actual behavior
mpv segfaults

### Log file

see attached logfile:
[hwdec-segfault-log.txt](https://github.com/mpv-player/mpv/files/7181774/hwdec-segfault-log.txt)

gdb back trace:

(gdb) run --no-config --vo=gpu --hwdec=auto '/mnt/nfs/TV/Parks.and.Recreation.S03E05.Media.Blitz.mkv' 
Starting program: /usr/local/bin/mpv --no-config --vo=gpu --hwdec=auto '/mnt/nfs/TV/Parks.and.Recreation.S03E05.Media.Blitz.mkv'
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7fff99c3b000 (LWP 4174570)]
[New Thread 0x7fff9943a000 (LWP 4174571)]
[New Thread 0x7fff98c39000 (LWP 4174572)]
[New Thread 0x7fff93fff000 (LWP 4174573)]
[New Thread 0x7fff937fe000 (LWP 4174574)]
[New Thread 0x7fff92ffd000 (LWP 4174575)]
[Thread 0x7fff92ffd000 (LWP 4174575) exited]
[New Thread 0x7fff92ffd000 (LWP 4174576)]
[Thread 0x7fff92ffd000 (LWP 4174576) exited]
[New Thread 0x7fff92ffd000 (LWP 4174577)]
[New Thread 0x7fff92488000 (LWP 4174578)]
 (+) Video --vid=1 (*) (hevc 1920x1080 23.976fps)
 (+) Audio --aid=1 --alang=eng (aac 6ch 48000Hz)
     Subs  --sid=1 --slang=eng (subrip)
[New Thread 0x7fff91c87000 (LWP 4174579)]

Thread 1 "mpv" received signal SIGSEGV, Segmentation fault.
0x00007ffff4b7c515 in __strlen_avx2 () from /usr/lib/libc.so.6
(gdb) bt full
#0  0x00007ffff4b7c515 in __strlen_avx2 () at /usr/lib/libc.so.6
#1  0x00007ffff4a86ada in printf_positional () at /usr/lib/libc.so.6
#2  0x00007ffff4a8733d in __vfprintf_internal () at /usr/lib/libc.so.6
#3  0x00007ffff4a99f8a in __vsnprintf_internal () at /usr/lib/libc.so.6
#4  0x00007ffff4a74396 in snprintf () at /usr/lib/libc.so.6
#5  0x0000555555647074 in add_all_hwdec_methods (infos=infos@entry=0x7fffffffe180, num_infos=num_infos@entry=0x7fffffffe17c)
    at ../video/decode/vd_lavc.c:392
        info = 
          {name = '\000' <repeats 63 times>, method_name = '\000' <repeats 23 times>, codec = 0x7ffff7458fa0, lavc_device = AV_HWDEVICE_TYPE_NONE, copying = true, pix_fmt = AV_PIX_FMT_NONE, use_hw_frames = false, use_hw_device = false, flags = 0, auto_pos = 0, rank = 0}
        info_template = 
          {name = '\000' <repeats 63 times>, method_name = '\000' <repeats 23 times>, codec = 0x7ffff7458fa0, lavc_device = AV_HWDEVICE_TYPE_NONE, copying = false, pix_fmt = AV_PIX_FMT_NONE, use_hw_frames = false, use_hw_device = false, flags = 0, auto_pos = 0, rank = 0}
        wrapper = 0x2000000006 <error: Cannot access memory at address 0x2000000006>
        found_any = false
        codec = 0x7ffff7458fa0
        iter = 0x131
        __PRETTY_FUNCTION__ = "add_all_hwdec_methods"
#6  0x0000555555648860 in select_and_set_hwdec (vd=vd@entry=0x555556a22e00) at ../video/decode/vd_lavc.c:469
        hwdecs = 0x555556a36430
        num_hwdecs = 12
        ctx = 0x555556a35de0
        codec = 0x5555556affd4 "hevc"
        opt = {start = <optimized out>, len = <optimized out>}
        hwdec_requested = <optimized out>
        hwdec_auto_all = <optimized out>
        hwdec_auto_safe = <optimized out>
        hwdec_auto_copy = <optimized out>
        hwdec_auto = <optimized out>
#7  0x0000555555648f6a in reinit (vd=0x555556a22e00) at ../video/decode/vd_lavc.c:579
        ctx = 0x555556a35de0
        use_hwdec = <optimized out>
        vd = 0x555556a22e00
        ctx = 0x555556a35de0
        info = <optimized out>
#8  create (parent=<optimized out>, codec=0x7fff7c021130, decoder=<optimized out>) at ../video/decode/vd_lavc.c:1280
        vd = 0x555556a22e00
        ctx = 0x555556a35de0
        info = <optimized out>
#9  0x00005555555d951c in reinit_decoder (p=0x555556a1f0b0) at ../filters/f_decoder_wrapper.c:453
        sel = 0x555556a3ab60
        n = 0
        driver = <optimized out>
        list = 0x555556a3a680
        user_list = <optimized out>
        fallback = <optimized out>
        p = 0x555556a1f0b0
        res = <optimized out>
#10 mp_decoder_wrapper_reinit (d=<optimized out>) at ../filters/f_decoder_wrapper.c:481
        p = 0x555556a1f0b0
        res = <optimized out>
--Type <RET> for more, q to quit, c to continue without paging--
#11 0x000055555562f6ea in init_video_decoder (mpctx=mpctx@entry=0x5555569874e0, track=track@entry=0x5555569b6f10)
    at ../player/video.c:189
        __PRETTY_FUNCTION__ = "init_video_decoder"
        parent = <optimized out>
#12 0x000055555562f82f in reinit_video_chain_src (mpctx=0x5555569874e0, track=0x5555569b6f10) at ../player/video.c:258
        __PRETTY_FUNCTION__ = "reinit_video_chain_src"
        vo_c = 0x555556a17430
#13 0x0000555555621f3b in play_current_file (mpctx=0x5555569874e0) at ../player/loadfile.c:1605
        opts = 0x55555698ed10
        playback_start = -1e+100
        __PRETTY_FUNCTION__ = "play_current_file"
        start_event = {playlist_entry_id = 1}
        end_event = {reason = 0, error = 0, playlist_entry_id = 1, playlist_insert_id = 0, playlist_insert_num_entries = 0}
        play_start_pts = <optimized out>
        nothing_played = <optimized out>
#14 0x0000555555623179 in mp_play_files (mpctx=mpctx@entry=0x5555569874e0) at ../player/loadfile.c:1864
        new_entry = <optimized out>
#15 0x0000555555627158 in mpv_main (argc=<optimized out>, argv=<optimized out>) at ../player/main.c:430
        mpctx = 0x5555569874e0
        options = <optimized out>
        r = 0
        rc = <optimized out>
        reason = <optimized out>
#16 0x00007ffff4a43b25 in __libc_start_main () at /usr/lib/libc.so.6
#17 0x000055555559987e in _start ()

**Comment by tantei3** [comment]
This issue is because of the following lines from the log file:
```
[   0.003][v][cplayer] mpv 0.33.0-270-g24357cb7b5 Copyright © 2000-2020 mpv/MPlayer/mplayer2 projects
[   0.003][v][cplayer]  built on Wed Aug  4 11:06:17 EDT 2021
[   0.003][v][cplayer] FFmpeg library versions:
[   0.003][v][cplayer]    libavutil       57.3.100 (runtime 57.5.100)
[   0.003][v][cplayer]    libavcodec      59.4.100 (runtime 59.7.103)
[   0.003][v][cplayer]    libavformat     59.4.101 (runtime 59.5.100)
[   0.003][v][cplayer]    libswscale      6.0.100 (runtime 6.1.100)
[   0.003][v][cplayer]    libavfilter     8.1.103 (runtime 8.9.100)
[   0.003][v][cplayer]    libswresample   4.0.100
[   0.003][v][cplayer] FFmpeg version: N-103654-g229e949c8e
```
Looks like your mpv was built against older version of ffmpeg (libavcodec 59.4.100) and at runtime there is a different version of ffmpeg shared library on the system (libavcodec 59.7.103). Rebuilding against the newer ffmpeg version should avoid this segfault.

**Comment by afarley** [comment]
So bizarre. You are right. I compile and recompile both ffmpeg and mpv almost daily, so this wasn't a concern. After several recompilations and several uninstall re-installs the problems isn't resolved (even the stock ffmpeg 4.4 and mpv 33.1 are complaining about weird crap).  It isn't at all clear on my end why this is happening as my other systems don't have this issue. This probably isn't a bug, so I apologize for submitting.
